scripts/fuzz_opt.py

```diff
@@ -1753,6 +1753,19 @@ class PreserveImportsExports(TestCaseHandler):
     frequency = 0.1
 
     def handle(self, wasm):
+        # We will later verify that no imports or exports changed, by comparing
+        # to the unprocessed original text.
+        original = run([in_bin('wasm-opt'), wasm] + FEATURE_OPTS + ['--print'])
+
+        # We cannot run if the module has (ref exn) in globals (because we have
+        # no way to generate an exn in a non-function context). The fuzzer is
+        # careful not to emit that in testcases, but after the optimizer runs,
+        # we may end up with struct fields getting refined to that.
```

Comment:
What's the connection between struct fields having type `(ref exn)` and globals having type `(ref exn)`?

Reply:
If a struct has a field `(ref exn)` and we pick that type for a global, we will end up doing `struct.new` with a list of initial values of the fields, one of whom will be `(ref exn)`.

Reply:
And do we always construct global structs by first constructing separate globals for each of their fields? Otherwise the `(ref exn)` type would not actually appear in the list of global definitions.

This all seems rather brittle :/

Reply:
No, we do not create separate globals for each field. The problem is not in a global having this type, the problem is trying to create such a value. E.g.
```wat
(type (struct $struct_with_exn (field i32) (field (ref exn)) (field f64)))

(global $struct_with_exn (struct.new $struct_with_exn
  (i32.const 42)
  .. what do we put here for (ref exn)? ..
  (f64.const 3.14159)
))
```
We have no way to create such a value in the global scope. In a function, we use the trick of `(try (throw))` i.e. we throw and catch, giving us a `(ref exn)`.

Reply:
Unless I'm missing something, that example won't be caught by the code below, though.

Reply:
Oh, good point! :laughing: That should say "struct", not "global"... I'll fix it.

It is really hard to test this stuff, unfortunately. This is in response to a random fuzzer input, and if we added such inputs to the test suite, they'd get out of date quickly, with no good way to find a replacement that happens to hit the same issue...